### PR TITLE
Port hardsubx utility

### PR DESCRIPTION
**Pull request by shashwat1002**
**In raising this pull request, I confirm the following (please check boxes):**

- [ ] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [ ] I have checked that another pull request for this purpose does not exist.
- [ ] I have considered, and confirmed that this submission will be valuable to others.
- [ ] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [ ] I give this submission freely, and claim no ownership to its content.
- [ ] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

- [] I have never used CCExtractor.
- [] I have used CCExtractor just a couple of times.
- [] I absolutely love CCExtractor, but have not contributed previously.
- [X] I am an active contributor to CCExtractor.

---

Ported `hardsubx_utility` 

added pure rust function for levenstein edit distance
added a wrapper for compatibility with C code. The wrapper won't be necessary when the downstream modules are ported.

PS: some methods in the original C code were never used anywhere and therefore they were not ported

CC: @PunitLodha

### CODE REVIEW

**Comment by shashwat1002** [comment]
@PunitLodha  it seems ffmpeg itself is not being built properly.

File: src/rust/build.rs [hunk 1]
```diff
@@ -42,5 +59,14 @@ const ALLOWLIST_FUNCTIONS: &[&str] = &[
     "printdata",
     "writercwtdata",
 ];
+
+#[cfg(feature = "hardsubx_ocr")]
+const ALLOWLIST_FUNCTIONS_OCR_HARDSUBX: &[&str] =
+    &["edit_distance", "convert_pts_to_.*", "av_rescale_q"];
+
```
Comment (PunitLodha) [inline src/rust/build.rs:66]: Could you change this and other lists to something like, 
``` rust
let mut allowlist_functions = vec![
    ".*(?i)_?dtvcc_.*",
    "get_visible_.*",
    "get_fts",
    "printdata",
    "writercwtdata",
];

#[cfg(feature = "hardsubx_ocr")]
allowlist_functions.extend_from_slice(&["edit_distance", "convert_pts_to_.*", "av_rescale_q"]);

```

Then there will be no need for another loop just to add hardsubx stuff

File: src/rust/build.rs [hunk 2]
```diff
@@ -18,10 +25,20 @@ fn main() {
         builder = builder.allowlist_function(fn_name);
     }
 
+    #[cfg(feature = "hardsubx_ocr")]
+    for fn_name in ALLOWLIST_FUNCTIONS_OCR_HARDSUBX {
+        builder = builder.allowlist_function(fn_name);
+    }
+
```
Comment (PunitLodha) [inline src/rust/build.rs:32]: No need for this, if you change the variables accordingly

File: src/rust/build.rs [hunk 3]
```diff
@@ -18,10 +25,20 @@ fn main() {
...
     for rust_enum in RUSTIFIED_ENUMS {
         builder = builder.rustified_enum(rust_enum);
     }
 
+    #[cfg(feature = "hardsubx_ocr")]
+    for rust_enum in ALLOWEDLIST_TYPES_OCR_HARDSUBX {
+        builder = builder.rustified_enum(rust_enum);
+    }
+
```
Comment (PunitLodha) [inline src/rust/build.rs:41]: Same as above

File: src/rust/src/hardsubx/utility.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
+// use crate::{
+//     bindings::*
+// };
+
+#[cfg(feature = "hardsubx_ocr")]
```
Comment (PunitLodha) [inline src/rust/src/hardsubx/utility.rs:5]: You have already added `#[cfg(feature = "hardsubx_ocr")` on `mod hardsubx`. Is it needed to be added to all functions in this file?

File: src/rust/src/hardsubx/utility.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+fn _edit_distance_rec(
+    word1: &[u8],
+    word2: &[u8],
+    i: usize,
+    j: usize,
+    dp_array: &mut Vec<Vec<i32>>,
+) -> i32 {
+    // Recursive implementation with DP of Levenshtein distance
```
Comment (PunitLodha) [inline src/rust/src/hardsubx/utility.rs:43]: @cfsmp3 could you take a look, if recursion is fine here?

File: src/rust/src/lib.rs [hunk 6]
```diff
@@ -13,7 +13,10 @@ pub mod bindings {
     include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
 }
 pub mod decoder;
+
```
Comment (PunitLodha) [inline src/rust/src/lib.rs:16]: Remove blank line

File: src/rust/src/lib.rs [hunk 7]
```diff
@@ -13,7 +13,10 @@ pub mod bindings {
     include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
 }
 pub mod decoder;
+
+#[cfg(feature = "hardsubx_ocr")]
 pub mod hardsubx;
+
```
Comment (PunitLodha) [inline src/rust/src/lib.rs:19]: Remove blank line. Let's keep the module definitions combined

**Review by PunitLodha — CHANGES_REQUESTED** [review]

File: src/rust/build.rs (hunk 1 above)
Comment (shashwat1002) [reply src/rust/build.rs:66]: Would have to move the definitions inside the main function then. They won't be globals anymore. 

That's not an issue right?

Comment (shashwat1002) [reply src/rust/build.rs]: Also @PunitLodha  this leads to mutability not being required if the flags are off and that leads to a warning and fails the build 

I will recommend keeping the original version

Comment (PunitLodha) [reply src/rust/build.rs]: That is due to the feature not being selected. Could you create a new PR and add the `--all-features` flag to the clippy command in the `format_rust` workflow

File: src/rust/src/hardsubx/utility.rs [hunk 8]
```diff
@@ -0,0 +1,91 @@
+// use crate::{
+//     bindings::*
+// };
```
Comment (PunitLodha) [inline src/rust/src/hardsubx/utility.rs:3]: Also, remove this

**Review by PunitLodha — CHANGES_REQUESTED** [review]
Just some formatting changes

**Comment by shashwat1002** [comment]
@PunitLodha please check

File: src/rust/src/hardsubx/utility.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    len1: ::std::os::raw::c_int,
+    len2: ::std::os::raw::c_int,
+) -> ::std::os::raw::c_int {
+    // The actual edit_distance function
+
+    let word1_string: &ffi::CStr;
+    let word2_string: &ffi::CStr;
+
+    word1_string = ffi::CStr::from_ptr(word1);
+    word2_string = ffi::CStr::from_ptr(word2);
```
[the remark this answers is not in the document]
Comment (PunitLodha) [reply src/rust/src/hardsubx/utility.rs]: @shashwat1002 I meant this, 
``` rust
    let word1_string = CStr::from_ptr(word1);
    let word2_string = CStr::from_ptr(word2);
```
There's no need here, to separate the declaration and initialization

[the remark this answers is not in the document]
Comment (shashwat1002) [reply src/rust/src/hardsubx/utility.rs]: My bad, changed it now @PunitLodha

**Review by PunitLodha — APPROVED** [review]

**Comment by PunitLodha** [comment]
Need to fix the formatting issues and builds. But it should another PR

**Pull request merged**